### PR TITLE
Shaders support for Android

### LINKED ISSUE #2060: Shaders don't work on Android (remove from menu or fix)
Shaders should be removed from the menu or fixed on Android (ogles2 supports them).

### PR DESCRIPTION
**Pull request by numberZero**
Fix #2060 
Tested on desktop with GLES2 emulation. Shaders seem to work well.

@stujones11 test results (on real Android):

config | result | comment
----|----|----
plain | works
generate normalmaps | works | no effect (as expected)
generate normalmaps, bump mapping | crashes
generate normalmaps, parallax occlusion | ? | syntax error fixed (?)
generate normalmaps, bump mapping, parallax occlusion | ? | syntax error fixed (?)
normalmaps included, bump mapping | works
normalmaps included, parallax occlusion | ? | syntax error fixed (?)
normalmaps included, bump mapping, parallax occlusion | ? | syntax error fixed (?)

Waving works, tone mapping works.

### CODE REVIEW

**Comment by stujones11** [comment]
Transparency is broken with basic shaders enabled, others seem to have no effect. Tested with ogles2 driver.

screen

**Comment by numberZero** [comment]
@stujones11 Make sure you compile with the ENABLE_GLES option. I thought it is mandatory on Android, but apparently it isn’t.

**Comment by numberZero** [comment]
On transparency, I have the same result for keyed textures (alpha blending, as on water, works well). Testing with #6971, should help, although may require changes in shaders.

**Comment by numberZero** [comment]
Also, logs are helpful usually ;)

**Comment by stujones11** [comment]
I will look at this more tomorrow, I'm pretty sure you are on the right path with this though, good job :)

**Comment by numberZero** [comment]
@stujones11 Thanks)
By the way, I just found out I managed to break the code right before pushing it to GH... Added a quickfix, should work)

**Comment by numberZero** [comment]
Something is wrong here... but shaders do work, including alpha testing and (fixed) alpha blending.
screenshot_20180701_004928
screenshot_20180701_005006
I feel GLES uses ABGR instead of ARGB, but screen capturing doesn’t account for that.

File: src/shader.cpp [hunk 1]
```diff
@@ -37,6 +37,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 #include "gamedef.h"
 #include "client/tile.h"
+#include "cmake_config.h"
```
Comment (stujones11) [inline src/shader.cpp]: This line breaks the android build as this file is only generated when you run cmake for the desktop build. I think this should be just `config.h`

**Comment by stujones11** [comment]
I have managed to get it working by `#if ENABLE_GLES` with `#ifdef __ANDROID__` in `shader.cpp` but this can probably fixed in android the build system.

Tone mapping and waving effects are working nicely, however, parallax occlusion has the same broken alpha effect as before and generate normal maps crashes the app on game start-up. I will see if I can obtain anything meaningful from the logcat.

screen

Note that I manually applied the last two commits if you wonder why the version hash is unchanged. Also notice the broken alpha on the wielditem, although I believe that's a separate issue.

**Comment by stujones11** [comment]
Logcat output with generate normal maps enabled https://gist.github.com/stujones11/fe4c4dea93ef3b8e08e87a326e60db74

**Comment by numberZero** [comment]
screenshot_20180708_141836
@stujones11

**Comment by numberZero** [comment]
Logcat output is messy, it contains many crash reports. In the latest one, it seems that `glGetString(GL_EXTENSIONS)` (in `Align2Npot2` at `client/tile.cpp:1104`) returned null. The only reason for that I can imagine is that OpenGL wasn’t initialized yet at that point.

> when setting bump mapping or generate normals independently it does not crash, although there is no effect (as expected I guess)

Bump mapping should have effect if normal maps are available, as in some texture packs. Parallax occlusion uses them as well.

> Parallax occlusion still has same broken alpha test.

Is there a log with parallax enabled?

> error C1008: undefined variable "varColor"

My bad, sorry.

**Comment by stujones11** [comment]
> The only reason for that I can imagine is that OpenGL wasn’t initialized yet at that point.

Surely not, the main menu, etc works just fine until you launch the game. I will try to obtain some better backtraces and more verbose logs. I am sure it will be something simple if it works under emulation.

This does look very promising, however, there are still some issues remaining to properly address #2060 IMO the shaders menu should be disabled for ogles1 since they can never work. It would also be nice if we could automatically select the most up-to-date driver, though I realize that is beyond the scope of this PR.

**Comment by stujones11** [comment]
@numberZero I got this with parallax enabled
```
ERROR[Main]: Irrlicht: Fragment shader compilation failed.
ERROR: 0:173: '/' :  wrong operand types  no operation '/' exists that takes a left-hand operand of type 'const float' and a right operand of type 'const int' (or there is no acceptable conversion)
ERROR: 0:173: '.0' : Syntax error:  syntax error
ERROR: 2 compilation errors.  No code generated.
```
Also bump mapping does appear to work when the normal map is included in a texture pack. The problem seems to be with the auto generation of the normal maps.

**Comment by numberZero** [comment]
Updated according to #7563 .

**Comment by numberZero** [comment]
I have no Android device, can anyone help? @arvego?
Here is [the build](https://yadi.sk/d/eamtdwgZ3aJWag), it should work but may not: I’ve never compiled anything for Android before. And shaders themselves may or may not work, depends. (Note that they are only available if you set `video_driver` to `ogles2` as GLES 1 doesn’t support them at all).

Also, there are [some other builds](https://yadi.sk/d/u6QX3rBl3aJVrE) as well (ca8ec468 is in master; 9d46cee9 is this PR before update&rebase).

**Comment by numberZero** [comment]
@stujones11 There are 3 outcomes possible currently:
1. This PR may be abandoned.
2. Someone may help me debug and improve this.
3. This PR may be merged as-is.

I certainly don’t like first option. Second one is possible in theory but requires someone’s else effort, preferably with some knowledge. And on third option... well, something is better than nothing; it is labeled as experimental, and doesn’t crashes on startup (that would prevent restoring settings), so that may be acceptable.

To fix this, I need to know where is the problem, at least. It may be in some shader, in C++ code, or maybe in your driver. Huge log mixed from many launches doesn’t help much.

P.S. There are 2 commits since last time you commented: 5fb1b8f (9d46cee before rebase) and b84a8e0.

**Comment by stujones11** [comment]
> There are 2 commits since last time you commented

The last commit I tested was 5fb1b8f which, iirc, did fix the syntax error. I though I had told you that but clearly I mustn't have. I was just about to test the latest commits, however, it looks someone has broken the android build again and I am not sure if I can simply revert the breaking commit.

**Comment by numberZero** [comment]
Actually, there **is** something relevant. Last line MT wrote is “Compiling high level shaders for nodes_shader,” that is `shader.cpp:762`, after which it exits *normally*. “Audio: Global Deinitialized” is written at program exit (by the destructor of `SoundManagerSingleton`).

**Comment by stujones11** [comment]
> Actually, there is something relevant

Well we already knew it was to do with the shaders but that is an interesting observation regarding the audio de-initialization. However, this is far from a clean exit. If I had to guess I'd say the app is trying to access somewhere it shouldn't and being killed at a much lower level.

**Comment by numberZero** [comment]
@stujones11 At hard exit, like SEGV or FPE, destructors aren’t called. The strange thing is that your debugger didn’t report the exit code which it should regardless of circumstances. Otherwise that looks like `exit(0)`.

**Comment by stujones11** [comment]
> Otherwise that looks like exit(0)

With `abort()` or `exit(0)` you always get the exit code. Fun fact, minetest **never** ends cleanly on android and never did :D

**Comment by stujones11** [comment]
> there is probably a bug in your system

I really don't think so but I guess we need someone else to prove this. It is not helping that the android build system is busted by a trivial line-ending conflict in the irrlicht patch files and nobody checks that massive PRs like 153fb211ac2342907eb766a79c1f41824f981ab5 even work on all supported platforms :P

**Comment by numberZero** [comment]
Tested on my friend’s phone. Works well with all shaders enabled, although the water looks strange. (I asked him to post a screenshot but he seemingly forgot that).

**Comment by stujones11** [comment]
> Works well with all shaders enabled

Screenshot with auto-generated bump-mapping enabled or it never happened :P

**Comment by paramat** [comment]
@sfan5 ?^

**Comment by stujones11** [comment]
Well that error message seems more to do with the server list than it has the shaders.

I have no idea about the water glitch, I am not seeing that, only my device crashes with normal generation enabled. It is a fairly generic SG Tab-A so I would guess that many others will have the same problem. This could be related to anything from android version to vendor specific drivers, I have no idea.

However, if this feature is off by default and labelled as experimental than I see no harm in it being merged. It may be that it works fine on many devices. The tone mapping and waving effect work fine for me, although I would never enable then since it easily halves my frame-rate.

**Comment by numberZero** [comment]
> android version

SG Tab-A has Android™ 5.0, regarding to [its page](https://www.samsung.com/us/mobile/tablets/all-other-tablets/samsung-galaxy-tab-a-8-0-16gb-wi-fi-smoky-titanium-sm-t350nzaaxar/). Latest Android™ is 8.0...

> vendor specific drivers

Most painful point. Driver bugs sit here and there all the time, yet can’t be found without access to the particular device in question.

**Off-topic:**
For an example of a vendor-specific bug, my BIOS reports incorrect, and not even nearly self-consistent, RAM layout. While it is expected to be the only reliable source of that data.*

> Edit: Perhaps the proto error is caused by the revert of 153fb21 ?

153fb21 changes auth, while the error there is that it can’t get the server list. Probably there was no internet connection at the moment (the world is SP).

P.S. MT GUI on Android is unusable.

**Comment by stujones11** [comment]
> MT GUI on Android is unusable.

As well as off-topic this is not very constructive. I will admit that it is far from perfect but it plays reasonably well on my device and accounts for likely over 75% of the userbase for online servers.

**Comment by numberZero** [comment]
> > MT GUI on Android is unusable.
>
> As well as off-topic this is not very constructive. I will admit that it is far from perfect but it plays reasonably well on my device.

Sorry, I should have posted a screenshot. But the device is not mine, so that’s not that easy. Anyway, that’s topic of #5054.

**Comment by numberZero** [comment]
@paramat This is not WIP.

**Comment by stujones11** [comment]
@numberZero Any chance you can rebase this (again). I would like to try it on my new phone.

**Comment by stujones11** [comment]
Well the good news is that it doesn't crash, however, the rendering is very glitchy.

Huawei honor 9 lite.

screen

**Comment by numberZero** [comment]
@stujones11 You’re probably right (I suppose you meant vertex shader; geometry one is [a different thing](https://www.khronos.org/opengl/wiki/Geometry_Shader)). To be sure, test with basic shaders only, please.

**Comment by stujones11** [comment]
> I suppose you meant vertex shader

Yeah, I those mixed up. It is certainly much better with only the basic shaders but still not perfect.

screen

**Comment by numberZero** [comment]
I suppose the remained gaps are those seen as occasional transparent pixels on PC. Try commenting out lines 915–922 in `mapblock_mesh.cpp`:
```c++
			if (next_makes_face == makes_face
					&& next_p_corrected == p_corrected + translate_dir
					&& next_face_dir_corrected == face_dir_corrected
					&& memcmp(next_lights, lights, ARRLEN(lights) * sizeof(u16)) == 0
					&& next_tile.isTileable(tile)) {
				next_is_different = false;
				continuous_tiles_count++;
			}
```
You will have more faces but proper adjacency in the mesh. That should close the gaps.

**Comment by stujones11** [comment]
> Try commenting out lines 915–922 

That does indeed close those gaps :+1: However, there is still some 'flickering' of similar glimpses of the sky as you move and look around, consequently, I was unable to capture anything in a screenshot.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest/pull/7514#issuecomment-414079942 , author is asking for help with this, it is not neglected by the author, so i think keep it open?

File: src/shader.cpp [hunk 2]
```diff
@@ -594,7 +621,41 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
+			"attribute vec4 inTexCoord0;\n" \
+/* 			"attribute vec4 inTexCoord1;\n" */\
+			"attribute vec4 inVertexTangent;\n" \
+			"attribute vec4 inVertexBinormal;\n" \
+			"\n" \
+			"#define gl_ModelViewMatrix mWorldView\n" \
+			"#define gl_NormalMatrix mNormal\n" \
+			"#define gl_Vertex inVertexPosition\n" \
+			"#define gl_Normal inVertexNormal\n" \
+			"#define gl_Color inVertexColor\n" \
```
Comment (sfan5) [inline src/shader.cpp]: all the `\` are not necessary here, btw

**Comment by paramat** [comment]
No response to requested changes for 6 months, which would mean closing.

However, i expect the author is no longer active in MT, and was asking for help, so this needs adoption if we want this contribution.

**Comment by numberZero** [comment]
@paramat I still read my email) I may even come back and write some code. But only if there is someone in MT who can decide what is needed to be written. I see my own >1 year old PRs without even a “Rebase needed” label. Just lost in time.

So I will now comment in each PR describing its status. Then it’s up to the core devs what to do.

**Comment by numberZero** [comment]
Beginning with this particular PR. I can’t finish it currently as I still have no Android device, and no experience with Android at all. So if there is anyone who can take this PR and finish it that would be good. Otherwise it’s probably better to close it until a better time.

**Comment by paramat** [comment]
I am closing all PRs labelled with 'Adoption? (PR abandoned)' because there is no sense in keeping an abandoned PR open, the label is present and allows filtering for abandoned PRs if anyone wants to adopt one.

**Comment by numberZero** [comment]
I finally got an Android device. Rebased&updated this PR (the branch it originates from, at least). Works for me™

**Comment by rubenwardy** [comment]
Unfortunately, this pr can't be reopened as the branch was forced pushed to. You will need to make a new branch and a new pr

**Pull request closed without merging**